### DISCUSSION TITLE
Installation with Poetry?

### DISCUSSION THREAD

**Discussion by ukerzel** [pyg-team/pytorch_geometric#4563] [Q&A]
Does anyone have experience installing pytorch geometric using poetry?
Poetry has some nice features but I can't quite figure out how to install torch-scatter and torch-sparese as they require the installation of specific versions obtained from a dedicated webpage?

Many thanks in advance

**Comment by rusty1s** [comment]
Can't https://data.pyg.org/whl/torch-1.11.0+cpu.html be used as an index for poetry? Reading through https://github.com/python-poetry/poetry/issues/1391 at the moment.

**Reply by ukerzel** [reply]
Hello, 

many thanks for the quick reply.

Hm. I added the following according to https://python-poetry.org/docs/repositories/

[[tool.poetry.source]]
name = "PyG"
url = "https://data.pyg.org/whl/torch-1.11.0+cpu.html"

to the configuration file pyproject.toml but that then resulted in : 

$ poetry add torch-scatter

  RepositoryError
 403 Client Error: Forbidden for url: https://data.pyg.org/whl/torch-1.11.0+cpu.html/torch-scatter/

trying instead:
$  poetry config repositories.PyG https://data.pyg.org/whl/torch-1.11.0+cpu.html

it seems to work - though it takes a long time (like >>30min for torch-scatter), it seems it needs to compile packages.

Does using the URL with "CPU" in it means that the CPU only versions are to be installed?

**Reply by rusty1s** [reply]
Got it. It looks like we need to support sub-directories for each individual extension package as well:
```
https://data.pyg.org/whl/torch-${TORCH}+${CUDA}/${PACKAGE}
```
Let me see if I can fix this easily :)

If it takes a long time to install, then this indeed means you are installing from source. Defining `cpu` means that you are using the CPU version. For CUDA, you may want to replace this with either `cu102`, `cu113` or `cu115`.

**Reply by ukerzel** [reply]
Great, many thanks for your help
